pkg/apis/sources/v1alpha1/rabbitmq_validation.go

```diff
@@ -19,40 +19,37 @@ package v1alpha1
 import (
 	"context"
 
-	"github.com/google/go-cmp/cmp/cmpopts"
-
 	"knative.dev/pkg/apis"
 	"knative.dev/pkg/kmp"
 )
 
 func (current *RabbitmqSource) Validate(ctx context.Context) *apis.FieldError {
-	if !current.Spec.Predeclared && current.Spec.ExchangeConfig.Name == "" {
-		return &apis.FieldError{
-			Message: "Name of exchange must be provided when spec.predeclared is false",
-			Paths:   []string{"spec", "exchangeConfig", "name"},
+	if current.Spec.RabbitmqResourcesConfig == nil {
+		return apis.ErrMissingField("rabbitmqResourcesConfig").ViaField("spec")
+	} else {
+		if !current.Spec.RabbitmqResourcesConfig.Predeclared && current.Spec.RabbitmqResourcesConfig.ExchangeName == "" {
+			return &apis.FieldError{
+				Message: "Name of exchange must be provided when spec.predeclared is false",
+				Paths:   []string{"spec", "rabbitmqResourcesConfig", "name"},
+			}
 		}
 	}
 
-	var errs *apis.FieldError
 	if current.Spec.RabbitmqClusterReference == nil {
 		return apis.ErrMissingField("rabbitmqClusterReference").ViaField("spec")
 	} else {
 		if current.Spec.RabbitmqClusterReference.Name == "" {
-			return errs.Also(apis.ErrMissingField("name").ViaField("rabbitmqClusterReference").ViaField("spec"))
-		}
-		if current.Spec.RabbitmqClusterReference.Namespace == "" {
-			return errs.Also(apis.ErrMissingField("namespace").ViaField("rabbitmqClusterReference").ViaField("spec"))
-		}
-		if current.Spec.RabbitmqClusterReference.ConnectionSecret != nil {
+			if current.Spec.RabbitmqClusterReference.ConnectionSecret == nil {
+				return apis.ErrMissingField("name").ViaField("rabbitmqClusterReference").ViaField("spec")
+			}
+		} else if current.Spec.RabbitmqClusterReference.ConnectionSecret != nil {
 			return apis.ErrDisallowedFields("connectionSecret").ViaField("rabbitmqClusterReference").ViaField("spec")
 		}
 	}
 
 	if apis.IsInUpdate(ctx) {
```

Comment:
We'd want this block to be the first block. So if the user goes from having the spec populated to being empty for example, we'd want to error and say spec is immutable. Right now, the webhook would complain about `RabbitmqResourcesConfig` is missing which can mislead the user into thinking the spec is mutable.

Reply:
just reiterating that the update bit should be the first thing we check for.